apps/agentic-chat/src/components/assistant-ui/ApproveUI.tsx

```diff
@@ -1,51 +1,63 @@
-import { makeAssistantToolUI } from '@assistant-ui/react';
+import {
+  makeAssistantToolUI,
+  ToolCallContentPartComponent,
+} from '@assistant-ui/react';
 import { BadgeCheck } from 'lucide-react';
 import { TextShimmer } from '../TextShimmer';
 import { CollapsableDetails } from './CollapsableDetails';
+import { useAssetsStore } from '../../stores/assets';
+import { ApproveParams, ApproveResult } from '../../tools/approve';
 
-export type ApproveArgs = {
-  token: string;
-  spender: string;
-  amountCryptoPrecision: string;
-  chainId: number;
-  decimals: number;
-};
+const Icon = BadgeCheck;
 
-export type ApproveResult = string; // transaction hash
+const ApproveUiContent: ToolCallContentPartComponent<
+  {
+    assetId: string;
+    spender: string;
+    amountCryptoPrecision: string;
+  },
+  string
+> = ({ status, result, args, isError, toolName }) => {
+  const assetsStore = useAssetsStore();
+  const asset = assetsStore.assetsById[args.assetId];
 
-const Icon = BadgeCheck;
+  switch (status.type) {
+    case 'running':
+    case 'requires-action':
+    case 'incomplete': {
+      if (!(args.assetId && asset))
```

Comment:
with zod for validation, does it automatically fail and respond accordingly if supplied args don't meet the schema? it feels like we should always have `assetId` on args, so we could use asset reference is missing `asset` instead, then feels like spender should be included as well.

Reply:
@kaladinlight to avoid repeating this over across tools, and to bring type safety in, we could supercede assistant UI with either a strict empty object type with no additional properties (i.e `Record<string, never>`) or the params types, WDYT? 
Then we could simply check for `isEmpty()` instead across the board (I'd need to triple check that all args indeed arrive in one go while streaming, but from the debugging I've done earlier, looks like they do)

Reply:
20365c6f986c11db67a088d6b59b93f6a88ad2f6 gm